## Content

## 01. Import libraries
## 02. Import data
## 03. Data wrangling

## 01. Import libraries

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Create path

path = r'C:\Users\Frederick\Documents\07-12-2023 Instacart Basket Analysis'

## 02. Import data

In [3]:
# Import new customer data

df_cust = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [4]:
# Import orders_products with flags data

df_ords_prods = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_flags.pkl'))

## 03. Data Wrangling

In [5]:
# Checking first 5 rows of customer data

df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [6]:
# Checking last 5 rows of customer data

df_cust.tail()

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095

In [7]:
# Checking on basis statistics

df_cust.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

## The basic statistics looks alright. Maximum age is 81, minimum age is 18. Minimum income is 25903, maximum of 593901 seems a bit higher than the rest, can explore this a bit futher later. 

In [8]:
# Drop columns that are not relevant - customer's name and date join. 

df_cust.drop(['First Name', 'Surnam', 'date_joined'], axis = 1, inplace = True)

## Dropping customer's first name, last name and date joind as we shouldn't be needing this for the analysis. 

In [9]:
# Rename column 'fam_status' to be marital_status

df_cust.rename(columns = {'fam_status' : 'marital_status'}, inplace = True)

## Rename column fam_status to marital_status to make it clearer

In [10]:
# Checking first 5 rows of customer data after data wrangling

df_cust.head()

user_id  Gender       STATE  Age  n_dependants marital_status  income
0    26711  Female    Missouri   48             3        married  165665
1    33890  Female  New Mexico   36             0         single   59285
2    65803    Male       Idaho   35             2        married   99568
3   125935  Female        Iowa   40             0         single   42049
4   130797  Female    Maryland   26             1        married   40374

In [11]:
# Check for mixed type data

for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)

C:\Users\Frederick\AppData\Local\Temp\ipykernel_12088\4127571237.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Frederick\AppData\Local\Temp\ipykernel_12088\4127571237.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Frederick\AppData\Local\Temp\ipykernel_12088\4127571237.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Frederick\AppData\Local\Temp\ipykernel_12088\4127571237.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Fre

## No mix type data identified.

In [12]:
# Finding missing values in customer dataframe

df_cust.isnull().sum()

user_id           0
Gender            0
STATE             0
Age               0
n_dependants      0
marital_status    0
income            0
dtype: int64

## No missing value identified. 

In [13]:
# Finding duplicate record with same column values

df_dups = df_cust[df_cust.duplicated()]

In [14]:
# Check output for duplicates

df_dups

Empty DataFrame
Columns: [user_id, Gender, STATE, Age, n_dependants, marital_status, income]
Index: []

## No duplicates identified

In [15]:
# Check 'user_id' type in customer dataframe

df_cust['user_id'].dtype

dtype('int64')

In [16]:
# Check 'user_id' in orders_products dataframe

df_ords_prods['user_id'].dtype

dtype('int64')

## The 'user_id' column that can join the customer dataframe and orders_products dataframe are of the same type - int64.

In [17]:
# Checking counts on marital status to ensure data is consistent

df_cust['marital_status'].value_counts(dropna = False)

marital_status
married                             144906
single                               33962
divorced/widowed                     17640
living with parents and siblings      9701
Name: count, dtype: int64

In [18]:
# Checking shape for customer dataframe

df_cust.shape

(206209, 7)

In [19]:
# Checking shape for orders_products dataframe

df_ords_prods.shape

(32404859, 28)

In [20]:
# Remove the limit on display option on columns

pd.options.display.max_columns = None

In [21]:
df_ords_prods.head()

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  user_order_number  \
0             0             0   2539329        1                  1   
1             1             1   2398795        1                  2   
2             2             2    473747        1                  3   
3             3             3   2254736        1                  4   
4             4             4    431534        1                  5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                   2                  8                     NaN         196   
1                   3                  7                    15.0         196   
2                   3                 12                    21.0         196   
3                   4                  7                    29.0         196   
4                   4                 15                    28.0         196   

   add_to_cart_order  reordered _merge  Unnamed: 0_y product_name  aisle_id  \
0                  1          0   both           195         Soda        77   
1                  1          1   both           195         Soda        77   
2                  1          1   both           195         Soda        77   
3                  1          1   both           195         Soda        77   
4                  1          1   both           195         Soda        77   

   department_id  prices exists     busiest_day    price_range_loc  \
0              7     9.0   both  Regularly busy  Mid-range product   
1              7     9.0   both  Regularly busy  Mid-range product   
2              7     9.0   both  Regularly busy  Mid-range product   
3              7     9.0   both      Least busy  Mid-range product   
4              7     9.0   both      Least busy  Mid-range product   

   busiest_days busiest_period_of_day  max_order  loyalty_flag  avg_price  \
0  Regular days        Average orders         10  New customer   6.367797   
1  Slowest days        Average orders         10  New customer   6.367797   
2  Slowest days           Most orders         10  New customer   6.367797   
3  Slowest days        Average orders         10  New customer   6.367797   
4  Slowest days           Most orders         10  New customer   6.367797   

  spending_flag  median_days_since_prior_order         frequency_flag  
0   Low spender                           20.5  Non-frequent customer  
1   Low spender                           20.5  Non-frequent customer  
2   Low spender                           20.5  Non-frequent customer  
3   Low spender                           20.5  Non-frequent customer  
4   Low spender                           20.5  Non-frequent customer

In [22]:
# Drop columns that are not relevant in orders_products dataframe 

df_ords_prods.drop(['Unnamed: 0.1', 'Unnamed: 0_x', '_merge', 'Unnamed: 0_y', 'exists', 'add_to_cart_order'], axis = 1, inplace = True)

In [23]:
# Check df_ords_prods columns after droppiong irrelevant columsn

df_ords_prods.head()

order_id  user_id  user_order_number  orders_day_of_week  \
0   2539329        1                  1                   2   
1   2398795        1                  2                   3   
2    473747        1                  3                   3   
3   2254736        1                  4                   4   
4    431534        1                  5                   4   

   order_hour_of_day  days_since_prior_order  product_id  reordered  \
0                  8                     NaN         196          0   
1                  7                    15.0         196          1   
2                 12                    21.0         196          1   
3                  7                    29.0         196          1   
4                 15                    28.0         196          1   

  product_name  aisle_id  department_id  prices     busiest_day  \
0         Soda        77              7     9.0  Regularly busy   
1         Soda        77              7     9.0  Regularly busy   
2         Soda        77              7     9.0  Regularly busy   
3         Soda        77              7     9.0      Least busy   
4         Soda        77              7     9.0      Least busy   

     price_range_loc  busiest_days busiest_period_of_day  max_order  \
0  Mid-range product  Regular days        Average orders         10   
1  Mid-range product  Slowest days        Average orders         10   
2  Mid-range product  Slowest days           Most orders         10   
3  Mid-range product  Slowest days        Average orders         10   
4  Mid-range product  Slowest days           Most orders         10   

   loyalty_flag  avg_price spending_flag  median_days_since_prior_order  \
0  New customer   6.367797   Low spender                           20.5   
1  New customer   6.367797   Low spender                           20.5   
2  New customer   6.367797   Low spender                           20.5   
3  New customer   6.367797   Low spender                           20.5   
4  New customer   6.367797   Low spender                           20.5   

          frequency_flag  
0  Non-frequent customer  
1  Non-frequent customer  
2  Non-frequent customer  
3  Non-frequent customer  
4  Non-frequent customer

In [24]:
# Merging customer dataframe with orders_products dataframe with the default inner join, in a new df_merged dataframe

df_merged = df_ords_prods.merge(df_cust, on = 'user_id', indicator = True)

MemoryError: Unable to allocate 2.41 GiB for an array with shape (10, 32404859) and data type int64

In [ ]:
# Checking output of the merged customer & orders_products dataframes

df_merged.head()

In [ ]:
# Checking count of the newly merged dataframe

df_merged['_merge'].value_counts()

## By looking at the count, it appears that all records from the customer dataframe have been merged as the total count on the merged file matched the total count from the orders_products dataframe. 

In [ ]:
# Exporting merged data to prepared data folder

df_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_customers.pkl'))